src/VisualStudio/Core/Def/Implementation/SyncNamespaces/SyncNamespacesCommandHandler.cs

```diff
@@ -0,0 +1,153 @@
+                    && GetProjectsForHierarchy(projectHierarchy)
+                        .All(project => project.Language.Equals(LanguageNames.CSharp, StringComparison.OrdinalIgnoreCase));
+            }
+            else
+            {
+                // Is a solution node. Do we contain any C# projects?
+                visible = _workspace.CurrentSolution.Projects
+                    .Any(project => project.Language.Equals(LanguageNames.CSharp, StringComparison.OrdinalIgnoreCase));
+            }
+
+            enabled = visible && !VisualStudioCommandHandlerHelpers.IsBuildActive();
+
+            if (command.Visible != visible)
+            {
+                command.Visible = visible;
+            }
+
+            if (command.Enabled != enabled)
+            {
+                command.Enabled = enabled;
+            }
+        }
+
+        private void OnSyncNamespacesForSelectedProject(object sender, EventArgs args)
+        {
+            if (VisualStudioCommandHandlerHelpers.TryGetSelectedProjectHierarchy(_serviceProvider, out var projectHierarchy))
+            {
+                var projects = GetProjectsForHierarchy(projectHierarchy);
+
+                SyncNamespaces(projects);
+            }
+            else
+            {
+                var projects = _workspace.CurrentSolution.Projects
+                    .Where(project => project.Language.Equals(LanguageNames.CSharp, StringComparison.OrdinalIgnoreCase))
+                    .ToImmutableArray();
+
+                if (_workspace.CurrentSolution.ProjectIds.Count != projects.Length)
+                {
+                    // Some projects are not C# projects.
+                }
+
+                SyncNamespaces(projects);
+            }
+        }
+
+        private ImmutableArray<Project> GetProjectsForHierarchy(Shell.Interop.IVsHierarchy projectHierarchy)
+        {
+            var projectFilePath = projectHierarchy.TryGetProjectFilePath();
+
+            var solution = _workspace.CurrentSolution;
+            return solution.Projects
+                .Where(project => project.FilePath?.Equals(projectFilePath, StringComparison.OrdinalIgnoreCase) == true)
+                .ToImmutableArrayOrEmpty();
+        }
+
+        private void SyncNamespaces(ImmutableArray<Project> projects)
+        {
+            if (projects.IsEmpty)
+            {
+                return;
+            }
+
+            var syncService = projects[0].GetRequiredLanguageService<ISyncNamespacesService>();
+
+            Solution? solution = null;
+            var status = _threadOperationExecutor.Execute(ServicesVSResources.Sync_Namespaces, ServicesVSResources.Updating_namspaces, allowCancellation: true, showProgress: true, (operationContext) =>
+            {
+                ThreadHelper.JoinableTaskFactory.Run(async () =>
+                {
+                    solution = await syncService.SyncNamespacesAsync(projects, operationContext.UserCancellationToken).ConfigureAwait(false);
```

Comment:
ConfigureAwait(false) says "ignore my current context and switch to the thread pool thread if I can" - it doesn't appear as if below is expectating that - doesn't TryApplyChanges need to be on the UI thread?

Reply:
It does not.

Reply:
But you are showing a messagebox, which is expecting to block the main message pump.

Reply:
In my manual testing there wasn't a noticeable issue, but I am happy to change this to be more correct.

Reply:
The message box being shown is outside of the `_threadOperationExecutor.Execute`, which should mean the work inside here is free threaded but blocks the UI thread synchronously. Everything outside of this will be on the calling (UI) thread